src/java.base/share/classes/java/lang/invoke/TypeConvertingMethodAdapter.java

```diff
@@ -25,34 +25,43 @@
 
 package java.lang.invoke;
 
-import jdk.internal.org.objectweb.asm.MethodVisitor;
-import jdk.internal.org.objectweb.asm.Opcodes;
-import jdk.internal.org.objectweb.asm.Type;
+import java.lang.constant.ClassDesc;
+import java.lang.constant.ConstantDescs;
+import java.lang.constant.MethodTypeDesc;
+import java.util.Map;
+import jdk.internal.classfile.Classfile;
+import jdk.internal.classfile.CodeBuilder;
+import jdk.internal.classfile.Opcode;
+import jdk.internal.classfile.TypeKind;
 import sun.invoke.util.BytecodeDescriptor;
 import sun.invoke.util.Wrapper;
 import static sun.invoke.util.Wrapper.*;
 
-class TypeConvertingMethodAdapter extends MethodVisitor {
-
-    TypeConvertingMethodAdapter(MethodVisitor mv) {
-        super(Opcodes.ASM7, mv);
-    }
+class TypeConvertingMethodAdapter {
 
     private static final int NUM_WRAPPERS = Wrapper.COUNT;
 
-    private static final String NAME_OBJECT = "java/lang/Object";
+    private static final ClassDesc NAME_OBJECT = ConstantDescs.CD_Object;
     private static final String WRAPPER_PREFIX = "Ljava/lang/";
 
     // Same for all primitives; name of the boxing method
     private static final String NAME_BOX_METHOD = "valueOf";
 
     // Table of opcodes for widening primitive conversions; NOP = no conversion
-    private static final int[][] wideningOpcodes = new int[NUM_WRAPPERS][NUM_WRAPPERS];
+    private static final Opcode[][] wideningOpcodes = new Opcode[NUM_WRAPPERS][NUM_WRAPPERS];
```

Comment:
On an unrelated note, this can be annotated `@Stable` for arrays are mutable, so VM needs a hint to treat element access as constants.

Reply:
Yes, I've added @Stable annotation.